### PR TITLE
(WIP) Cleaned Flux2 Klein Implementation, with benchmarking done on v6 TPU

### PR DESCRIPTION
**Pull request by amepas**
(WIP) - this will be updated with multi-chip latency and support for Flux2 Klein 9B!

Draft PR for the Flux2 Klein model. Includes a custom implementation of the Qwen3-4B model for getting text embeddings. VAE Decoder, RoPE positional embedder, flow-matching step schedule are all re-used. Light modifications to transformer/attention blocks are used. 

Latency for batch-size 4 of 1024 by 1024 images (bfloat16): 
*   **Prompt Encoding (Qwen3)**: **`57.67 ms`** (1.58% of total)
*   **Denoising Loop (Flux 4 steps)**: **`3,181.20 ms`** (87.09% of total)
    *   *Per-Step Transformer Time*: **`795.30 ms`**
*   **VAE Decoding (VAE)**: **`413.77 ms`** (11.33% of total)
*   *Total*: 3.65 seconds

PR includes code for verifying accuracy of implementation. Sharding model is implemented but not tested.

Image generation is only supported so far.

### CODE REVIEW

File: src/maxdiffusion/tests/flux2klein/generate_flux2klein_e2e_test.py [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+    prepare_latent_image_ids,
+    pack_latents,
+    unpack_latents_with_ids,
+    unpatchify_latents,
+    encode_prompt_jax,
+)
+
+class GenerateFlux2KleinE2ETest(unittest.TestCase):
+
+    def test_end_to_end_parity_and_offloading(self):
```
Comment (entrpn) [inline src/maxdiffusion/tests/flux2klein/generate_flux2klein_e2e_test.py]: this test is very likely to fail in the github runner with the hardcoded values. We usually don't run e2e tests on the github runner, you can mark it so it doesn't run in the runner.

File: src/maxdiffusion/tests/flux2klein/generate_flux2klein_e2e_test.py [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+        """
+        Executes the entire generation pipeline (JAX Qwen3 -> JAX Flux -> JAX VAE)
+        under dynamic parameter offloading, validating mathematical parity at
+        every single stage against the golden PyTorch reference.
+        """
+        # Set highest precision for strict mathematical parity checks
+        jax.config.update("jax_default_matmul_precision", "highest")
```
Comment (entrpn) [inline src/maxdiffusion/tests/flux2klein/generate_flux2klein_e2e_test.py]: what is the reason for using highest here?

**Review by entrpn — CHANGES_REQUESTED** [review]

File: src/maxdiffusion/generate_flux2klein_9B.py
Comment (Perseus14) [inline src/maxdiffusion/generate_flux2klein_9B.py:1]: Do we need a separate generate file for 9B model? Can this be combined with the `generate_flux2klein.py` or even the `generate_flux.py`?

File: src/maxdiffusion/tests/flux2klein/benchmark_tpu_pure.py
Comment (Perseus14) [inline src/maxdiffusion/tests/flux2klein/benchmark_tpu_pure.py:1]: We don't need a benchmark code to be added to tests/ folder. We could benchmark the results and mention it in the PR or a doc but no need to add it to the repo

File: src/maxdiffusion/schedulers/scheduling_flow_match_flax.py [hunk 3]
```diff
@@ -93,6 +93,8 @@ def __init__(
       inverse_timesteps: bool = False,
       extra_one_step: bool = False,
       reverse_sigmas: bool = False,
+      use_dynamic_shifting: bool = False,
+      time_shift_type: str = "linear",
```
Comment (Perseus14) [inline src/maxdiffusion/schedulers/scheduling_flow_match_flax.py:97]: What are the possible `time_shift_types` here? Is moving to a Enum data type better?

File: src/maxdiffusion/generate_flux2klein.py
Comment (Perseus14) [inline src/maxdiffusion/generate_flux2klein.py:1]: I suspect you can refactor this file and move some of the functions to files under different folders like `pipeline`, `models/flux`,  `max_utils.py`

Refer to WAN model related files

**Review by Perseus14 — COMMENTED** [review]
Hey @amepas, I see this is WIP, added a few comments to help polish the PR

**Comment by amepas** [comment]
Closing this pull request because PRs 439 and 440 contain the updated code to incorporate the feedback!
